pkg/p2p/libp2p/libp2p.go

```diff
@@ -193,9 +194,17 @@ func New(ctx context.Context, o Options) (*Service, error) {
 		}
 
 		if exists := s.peers.addIfNotExists(stream.Conn(), i.Address); exists {
+			if err := stream.Close(); err != nil {
+				_ = stream.Reset()
+			}
+
 			return
 		}
 
+		if err := stream.Close(); err != nil {
+			_ = stream.Reset()
```

Comment:
if `Close` failed, won't `Reset` panic?

Reply:
See comment about about `Reset`. It will not panic, it looks like they use this as safeguard, if close fails to at least try to close other side of the stream and indicate something is wrong. Multiple `Close` and `Reset` should not panic I think. You can check `helpers.FullClose` implementation of `libp2p`, the original one, not ours. Ours is calling that one, so it should be easy to find. I don't like this as well, but it looks like we should do this on each `Close()` maybe.

Reply:
Then why don't you just use `helpers.FullClose`?

Reply:
See the explanation below, basically fullclose waits for other side to close the stream. I use it in Hive protocol and in handshake, as part of protocol to make sure that message is received, but here I don't think it is a right way to do it.